### ISSUE TITLE
[no-unused-vars] False positive when using decorators

### ISSUE THREAD

**Issue by njdancer** [typescript-eslint/typescript-eslint#557]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/indent": ["error", 2],
    "@typescript-eslint/member-delimiter-style": [
      "warn",
      {
        "multiline": {
          "delimiter": "none",
          "requireLast": true
        },
        "singleline": {
          "delimiter": "semi",
          "requireLast": false
        }
      }
    ],
    "@typescript-eslint/explicit-function-return-type": "off",
    "no-console": "warn"
  }
}
```

```TS
const Decorator = () => {}
const Entity = () => {}

@Decorator(returns => Entity)
class Test {
  public test() {
    return 1 + 1
  }
}
new Test()
```

**Expected Result**
Decorator parameter should not create unused var warning

**Actual Result**
```'returns' is defined but never used.```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.9.0` |
| `@typescript-eslint/parser`        | `1.9.0` |
| `TypeScript`                       | `3.4.5` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `12.3.1` |
| `yarn`                              | `1.16.0` |

**Comment by bradzacher** [comment]
Two things:
1 - It looks like there's more to your example than you've posted here, because none of the rules in your list should report that message. Could you please have a look at what the rulename reported with the error message is, and update with your config for that rule...

2 - If it's the rule I'm thinking of, then that should be a correct lint error for `@typescript-eslint/no-unused-vars`, because you have defined a parameter (`returns`) and then never used it.

**Comment by njdancer** [comment]
Yeah I should have included the configs I extend as well. I think the commit template may have just said rules so I took it literally. Don’t have it with me right now but can get you the full config if you need.

Yes it was no-unused-vars. I don’t know a lot about decorators and I’m struggling to find examples of this exact syntax but as I understand it, the returns parameter is being consumed by the decorator. I’m not even sure where returns would be in scope in that file. It is a parameter to decorator on the class. I’ve realised this example is slightly different to the original which was a decorator on the member function but it triggered the same warning. Essentially I’m doing something similar to this. https://typegraphql.ml/docs/getting-started.html

It seems to me the parser is interpreting that as a parameter declaration when it’s a value being passed as a parameter. I can’t say for certain the syntax is correct but if it isn’t I think this is not the error that should be reported. This syntax(not this example) is also interpreted correctly by tsc.

**Comment by bradzacher** [comment]
Reading and researching deeper into that library.
The library uses a lot of magic to work (looks like a cool lib though!).

The function being passed to the decorator is being consumed by the library via reflection.

However this doesn't change the fact that the function argument is actually being unused by the function body, i.e. the rule is reporting correctly.

---

This is use case is currently unsupported - we can't make assumptions about whether or not a decorator consumes a parameter via reflection or not.

However, I wouldn't be against someone adding an option to the rule to disable reporting for unused parameters within decorators.

**Comment by njdancer** [comment]
Thanks for the explanation. That makes sense now. I wasn't sure if it was just an arrow function or some kind of declarator specific syntax. Knowing that now, it is a pretty weird edge case for ESLint to know is correct so I'm satisfied continuing to use the disable comment on those lines. I may file an issue with typegraphql just for some feedback. The syntax is cool but I'm not sure its necessary to be able to pass a type reference into the decorator and in this instance it may cause someone to disable this rule which may not be a good trade off.

I'll leave this issue open based off your last comments but feel free to close it if you're happy with the rule the way it works.

**Comment by ark120202** [comment]
As far as I know TypeGraphQL docs use it only for documentation purposes, so just omitting it like `@Decorator(() => Entity)` should work.

**Comment by sgarner** [comment]
> As far as I know TypeGraphQL docs use it only for documentation purposes, so just omitting it like `@Decorator(() => Entity)` should work.

^ This is correct.

Another solution is to configure this eslint rule to ignore variables that match a pattern, for example vars starting with an underscore:

```
  "@typescript-eslint/no-unused-vars": ["warn", { "argsIgnorePattern": "^_" }]
```

Then you can use `_returns` as the function argument name and retain the readability benefits:

```ts
  @Decorator(_returns => Entity)
```

To better support this common use case (which is the recommended style for both TypeORM and TypeGraphQL decorators), perhaps there could be another option added to this rule to make it acceptable for an arrow function with an inline body and only a single argument to not use that argument?
